### PR TITLE
Centralized SR code generation

### PR DESCRIPTION
**Pull request by kvark**

### CODE REVIEW

**Comment by kvark** [comment]
@antiagainst I think it's looking more comprehensible now, saving about 40 LOC on the way :) Please take a look!

File: rspirv/sr/autogen_type_enum_check.rs [hunk 1]
```diff
@@ -17,7 +17,7 @@
 // DO NOT MODIFY!
 
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub(in crate::sr) enum TypeEnum {
+pub enum TypeEnum {
```
Comment (antiagainst) [inline rspirv/sr/autogen_type_enum_check.rs]: This should be `Type` internal; why making it entirely public?

File: rspirv/sr/types.rs [hunk 2]
```diff
@@ -29,7 +29,7 @@ pub struct Type {
 }
 
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub(in crate::sr) struct StructMember {
+pub struct StructMember {
```
Comment (antiagainst) [inline rspirv/sr/types.rs]: Same here. Why leaking this symbol to the public?

File: autogen/sr.rs [hunk 3]
```diff
@@ -114,166 +74,70 @@ fn get_type_fn_name(name: &str) -> String {
...
-pub fn gen_sr_type_check(grammar: &structs::Grammar) -> String {
-    // Collect all types and their parameters in the following format:
-    //   (type-name: &str, Vec<(param-name: quote::Ident, param-type: quote::Ident)>)
-    let cases: Vec<_> = grammar
-        .instructions
+pub struct GeneratedOperands {
+    pub decoration: String,
+}
+
+pub fn gen_sr_operands(grammar_operand_kinds: &[structs::OperandKind]) -> GeneratedOperands {
```
Comment (antiagainst) [inline autogen/sr.rs]: Can we name this function as `gen_sr_code_from_operand_kind_grammar` or something? "Operand" is already fairly overloaded; let's differentiate wherever possible.

File: autogen/sr.rs [hunk 4]
```diff
@@ -114,166 +74,70 @@ fn get_type_fn_name(name: &str) -> String {
     }
 }
 
-pub fn gen_sr_type_check(grammar: &structs::Grammar) -> String {
-    // Collect all types and their parameters in the following format:
-    //   (type-name: &str, Vec<(param-name: quote::Ident, param-type: quote::Ident)>)
-    let cases: Vec<_> = grammar
-        .instructions
+pub struct GeneratedOperands {
```
Comment (antiagainst) [inline autogen/sr.rs]: Name it as `CodeGeneratedFromOperandKindGrammar` or similar?

File: autogen/sr.rs [hunk 5]
```diff
@@ -114,166 +74,70 @@ fn get_type_fn_name(name: &str) -> String {
...
-            }
-        })
-        .collect();
-    let tokens = quote! {
-        impl Context {
-            #( #constructors )*
-        }
-    };
-    tokens.to_string()
+pub struct GeneratedInstructions {
```
Comment (antiagainst) [inline autogen/sr.rs]: Name it as `CodeGeneratedFromInstructionGrammar` or similar?

File: autogen/sr.rs [hunk 6]
```diff
@@ -114,166 +74,70 @@ fn get_type_fn_name(name: &str) -> String {
...
+    pub type_enums: String,
+    pub type_creation: String,
+    pub instruction_structs: String,
+    pub instruction_enums: String,
 }
 
-/// Generates instruction enums and standalone structs.
-pub fn gen_sr_instruction(grammar: &structs::Grammar) -> (String, String) {
-    let mut structs = Vec::new();
+pub fn gen_sr_instructions(grammar_instructions: &[structs::Instruction]) -> GeneratedInstructions {
```
Comment (antiagainst) [inline autogen/sr.rs]: Name it as `gen_sr_code_from_instruction_grammar` ?

File: autogen/sr.rs [hunk 7]
```diff
@@ -292,13 +156,89 @@ pub fn gen_sr_instruction(grammar: &structs::Grammar) -> (String, String) {
...
+                    })
+                    .collect();
+                let param_list = param_list.as_slice();
+                type_variants.push(if param_list.is_empty() {
+                    quote!{ #symbol }
+                } else {
+                    quote! { #symbol { #( #param_list ),* } }
+                });
+
+                if true {
```
Comment (antiagainst) [inline autogen/sr.rs]: Why the `if true` here?

**Comment by kvark** [comment]
Thanks for the review!

I renamed the types and functions as advised.
As for the public access, how else would the user be able to inspect SR of SPIR-V? My understanding was that we want to expose a nice type layer where users would match on their own and construct new data. That's also why I don't think `is_xxx` accessors are needed, by the way.

**Comment by antiagainst** [comment]
I actually view the `TypeEnum` field as internal to `Type`. All types should be uniqued into `Context` and created from `Context`. Hiding internals of `Type` and make its `new` method not public is for that purpose. There are two asks regarding `Type`: probing its properties, which is intended to be achieved via those `is_*` method, and constructing new types, which should be done via the `Context`. Without opening the internals of `Type` to the public, I don't quite see why making `TypeEnum` public would be useful? Instead of letting each call site to have a match construct to probe what kind a type is, why `is_*` not a nicer API that saves the boilerplate? 

Anyway, I think the change to publicize `TypeEnum` should be in another PR given that is not relevant to the PR here (which is trying to refactoring autogen) and should be discussed separately. Can you undo it to have this PR submitted first?

**Comment by antiagainst** [comment]
Typically I think smaller PRs dedicated to one task would be nice for review and revert if something goes wrong later.

**Comment by kvark** [comment]
Thanks for the review!
I agree the public qualifier change needs to be discussed separately (responded in #42). It is now reverted here. Are we good to go?

**Pull request merged**